File: server/src/tools/queryEngine.js
import { getDb } from "../data/db.js";

// Turns a granularity into the SQL expression used to bucket `date`.
// week buckets to the Monday that starts the week; month buckets to the 1st.
export function bucketExpr(granularity) {
  if (granularity === "week") return "date(date, '-' || ((strftime('%w', date) + 6) % 7) || ' days')";
  if (granularity === "month") return "strftime('%Y-%m-01', date)";
  return "date";
}

export class ToolInputError extends Error {}

function assertKnownValues(values, allowed, label) {
  if (!values || values.length === 0) return;
  const bad = values.filter((v) => !allowed.includes(v));
  if (bad.length > 0) {
    throw new ToolInputError(`Unknown ${label}: ${bad.join(", ")}. Valid values: ${allowed.join(", ")}`);
  }
}

/**
 * Builds and runs a parameterized, read-only SQL query grouping a value column
 * over time, optionally split by an entity dimension. This is the "backend
 * builds the query" layer: the model supplies structured, typed arguments —
 * never raw SQL — and this function is the only thing that touches the DB.
 */
export function runGroupedTimeSeries({
  table,
  entityCol,
  entityAllowed,
  entities,
  extraFilters = [], // [{col, allowed, values}]
  valueCol,
  aggregate = "SUM",
  startDate,
  endDate,
  granularity = "day",
  groupByEntity = true,
}) {
  assertKnownValues(entities, entityAllowed, entityCol);
  for (const f of extraFilters) assertKnownValues(f.values, f.allowed, f.col);

  const where = [];
  const params = {};
  if (startDate) {
    where.push("date >= @startDate");
    params.startDate = startDate;
  }
  if (endDate) {
    where.push("date <= @endDate");
    params.endDate = endDate;
  }
  if (entities && entities.length > 0) {
    const placeholders = entities.map((_, i) => `@entity${i}`).join(", ");
    entities.forEach((v, i) => (params[`entity${i}`] = v));
    where.push(`${entityCol} IN (${placeholders})`);
  }
  extraFilters.forEach((f, fi) => {
    if (f.values && f.values.length > 0) {
      const placeholders = f.values.map((_, i) => `@filter${fi}_${i}`).join(", ");
      f.values.forEach((v, i) => (params[`filter${fi}_${i}`] = v));
      where.push(`${f.col} IN (${placeholders})`);
    }
  });

  const bucket = bucketExpr(granularity);
  const selectEntity = groupByEntity ? `${entityCol} AS entity,` : "";
  const groupBy = groupByEntity ? `bucket, ${entityCol}` : "bucket";

  const sql = `
    SELECT ${bucket} AS bucket, ${selectEntity} ${aggregate}(${valueCol}) AS value
    FROM ${table}
    ${where.length ? "WHERE " + where.join(" AND ") : ""}
    GROUP BY ${groupBy}
    ORDER BY bucket ASC${groupByEntity ? `, ${entityCol} ASC` : ""}
  `;

  const rows = getDb().prepare(sql).all(params);
  return { sql: sql.trim(), rows };
}

/** Reshapes grouped rows [{bucket, entity, value}] into chart-ready wide rows:
 * [{ x: bucket, [entity]: value, ... }], plus the ordered list of series keys.
 */
export function pivotByEntity(rows, { fallbackSeriesName } = {}) {
  const byBucket = new Map();
  const seriesSet = [];
  for (const row of rows) {
    const key = row.entity ?? fallbackSeriesName ?? "value";
    if (!seriesSet.includes(key)) seriesSet.push(key);
    if (!byBucket.has(row.bucket)) byBucket.set(row.bucket, { x: row.bucket });
    byBucket.get(row.bucket)[key] = row.value;
  }
  return { data: Array.from(byBucket.values()), series: seriesSet };
}

File: server_py/tools/query_engine.py
"""Direct port of server/src/tools/queryEngine.js. Builds and runs a
parameterized, read-only SQL query grouping a value column over time,
optionally split by an entity dimension. This is the "backend builds the
query" layer: the model supplies structured, typed arguments -- never raw
SQL -- and this function is the only thing that touches the DB, via
whichever data source is active (SQLite stub data, or a real Snowflake
warehouse -- see data/data_source.py).
"""

from ..data.data_source import get_data_source
from ..data.result_sanity import check_result_sanity


class ToolInputError(Exception):
    pass


def bucket_expr(granularity: str) -> str:
    """Delegates to whichever data source is active -- the exact date-
    bucketing expression is dialect-specific (SQLite vs Snowflake date
    functions differ)."""
    return get_data_source().bucket_expr(granularity)


def _assert_known_values(values, allowed, label):
    if not values:
        return
    bad = [v for v in values if v not in allowed]
    if bad:
        raise ToolInputError(f"Unknown {label}: {', '.join(bad)}. Valid values: {', '.join(allowed)}")


def run_grouped_time_series(
    *,
    table: str,
    entity_col: str,
    entity_allowed: list[str],
    entities: list[str] | None = None,
    extra_filters: list[dict] | None = None,  # [{"col", "allowed", "values"}]
    value_col: str,
    aggregate: str = "SUM",
    start_date: str | None = None,
    end_date: str | None = None,
    granularity: str = "day",
    group_by_entity: bool = True,
) -> dict:
    extra_filters = extra_filters or []
    _assert_known_values(entities, entity_allowed, entity_col)
    for f in extra_filters:
        _assert_known_values(f.get("values"), f["allowed"], f["col"])

    where = []
    params: dict = {}
    if start_date:
        where.append("date >= @startDate")
        params["startDate"] = start_date
    if end_date:
        where.append("date <= @endDate")
        params["endDate"] = end_date
    if entities:
        placeholders = ", ".join(f"@entity{i}" for i in range(len(entities)))
        for i, v in enumerate(entities):
            params[f"entity{i}"] = v
        where.append(f"{entity_col} IN ({placeholders})")
    for fi, f in enumerate(extra_filters):
        values = f.get("values")
        if values:
            placeholders = ", ".join(f"@filter{fi}_{i}" for i in range(len(values)))
            for i, v in enumerate(values):
                params[f"filter{fi}_{i}"] = v
            where.append(f"{f['col']} IN ({placeholders})")

    data_source = get_data_source()
    bucket = data_source.bucket_expr(granularity)
    group_by = "bucket, entity" if group_by_entity else "bucket"

    # Two-stage aggregation. A table can carry a dimension the caller didn't
    # ask to split by (e.g. stablecoin_supply has both issuer and chain) --
    # rows sharing the same (date, entity) are components of one total and
    # must always be SUMmed first (a stablecoin's total supply is the sum of
    # its per-chain supplies, not their average). Only *then*, across the
    # distinct dates inside one time bucket, does the SUM-vs-AVG choice for
    # stock vs. flow metrics apply -- summing a stock like TVL across a week
    # of days would inflate it; averaging a flow like volume would deflate it.
    where_clause = f"WHERE {' AND '.join(where)}" if where else ""
    entity_select = f"{entity_col} AS entity," if group_by_entity else ""
    entity_group = f", {entity_col}" if group_by_entity else ""
    bucket_entity_select = "entity," if group_by_entity else ""
    order_entity = ", entity ASC" if group_by_entity else ""

    sql = f"""
        WITH daily AS (
          SELECT date, {entity_select} SUM({value_col}) AS value
          FROM {table}
          {where_clause}
          GROUP BY date{entity_group}
        )
        SELECT {bucket} AS bucket, {bucket_entity_select} {aggregate}(value) AS value
        FROM daily
        GROUP BY {group_by}
        ORDER BY bucket ASC{order_entity}
    """

    rows = data_source.run_query(sql, params)

    # The choke point every data tool's rows pass through regardless of
    # which adapter served them -- see data/result_sanity.py for why a
    # runaway row count fails loud (ToolInputError) while null-heavy or
    # stale data only annotates the result with a warning instead of
    # blocking it.
    try:
        warning = check_result_sanity(rows, value_col="value", end_date=end_date)
    except ValueError as err:
        raise ToolInputError(str(err)) from err

    return {"sql": sql.strip(), "rows": rows, "warning": warning}


def pivot_by_entity(rows: list[dict], fallback_series_name: str | None = None) -> dict:
    """Reshapes grouped rows [{bucket, entity, value}] into chart-ready wide
    rows: [{x: bucket, <entity>: value, ...}], plus the ordered series list."""
    by_bucket: dict[str, dict] = {}
    series_set: list[str] = []
    for row in rows:
        key = row.get("entity") or fallback_series_name or "value"
        if key not in series_set:
            series_set.append(key)
        bucket = row["bucket"]
        if bucket not in by_bucket:
            by_bucket[bucket] = {"x": bucket}
        by_bucket[bucket][key] = row["value"]
    return {"data": list(by_bucket.values()), "series": series_set}
